modules/gmake/tests/sln/sln_help_rule_tests.lua

```diff
@@ -0,0 +1,89 @@
+local gmake = require('gmake')
+local sln = gmake.sln
+
+local GmakeHelpRuleTests = test.declare('GmakeHelpRuleTests', 'gmake-sln', 'gmake')
+
+
+---
+-- Tests the help output with no projects or configurations.
+---
+function GmakeHelpRuleTests.NoProjectsNoConfigurations()
+	workspace('MyWorkspace', function ()
+		configurations({})
+	end)
+
+	local wks = gmake.buildDom().workspaces['MyWorkspace']
+
+	sln.helpRule(wks)
+
+	test.capture [[
+help:
+	@echo "Usage: make [config=name] [target]"
+	@echo ""
+	@echo "Targets:"
+	@echo "	all [default]"
+	@echo "	clean"
+	@echo "	help [Prints this message]"
+	@echo ""
+	@echo "For more information, see https://github.com/premake/premake-core/wiki"
+	]]
+end
+
+
+---
+-- Tests the help rule output with a single configuration and no projects.
+---
+function GmakeHelpRuleTests.NoProjectsSingleConfiguration()
+	workspace('MyWorkspace', function ()
+		configurations({ 'Debug' })
+	end)
+
+	local wks = gmake.buildDom().workspaces['MyWorkspace']
+
+	sln.helpRule(wks)
+
+	test.capture [[
+help:
+	@echo "Usage: make [config=name] [target]"
+	@echo ""
+	@echo "Configurations:"
+	@echo "	debug"
+	@echo ""
+	@echo "Targets:"
+	@echo "	all [default]"
+	@echo "	clean"
+	@echo "	help [Prints this message]"
+	@echo ""
+	@echo "For more information, see https://github.com/premake/premake-core/wiki"
+	]]
+end
+
+
+---
+-- Tests the help rule output with a single project and no configurations.
+---
+function GmakeHelpRuleTests.SingleProjectNoConfigurations()
+	workspace('MyWorkspace', function ()
+		configurations({})
+
+		project('MyProject', function ()
+		end)
+	end)
+
+	local wks = gmake.buildDom().workspaces['MyWorkspace']
+
+	sln.helpRule(wks)
+
+	test.capture [[
+help:
+	@echo "Usage: make [config=name] [target]"
+	@echo ""
+	@echo "Targets:"
+	@echo "	all [default]"
+	@echo "	clean"
+	@echo "	help [Prints this message]"
+	@echo "	MyProject"
+	@echo ""
+	@echo "For more information, see https://github.com/premake/premake-core/wiki"
```

Comment:
This will need to get changed in both v5 & v6, should be https://premake.github.io now.

Reply:
> I'm not sure how the gmake2 generator supported multiple jobs (`-j#`) properly compared to the gmake generator, but this rewrite needs to ensure it doesn't go backwards on supporting multiple jobs.



> > The `-j` flag is done on command line, not in a rule, but I definitely agree we don't want to break multi-core/multi-job builds.
> 
> I know that, but there was _something_ about the original gmake generator that caused build failures if you used the `-j` flag which the gmake2 generator fixed. My only concern about another rewrite is reintroducing this and creating a reason to rewrite the generator again.

Ohhh, following now :).  I'll keep an eye out for it and see if I can get it to break.

Reply:
Yeah, I'm super NOT thrilled to be rewriting this stuff again, TBH. But between the two different modules and all the various authors it's kind of a mess. To @samsinsane's point: definitely work against the Gmake2 module as much as possible, if you weren't already, as that's been getting the bulk of the recent fixes.